e2e/nomostest/testpredicates/predicates.go

```diff
@@ -250,36 +249,98 @@ func HasExactlyImage(containerName, expectImageName, expectImageTag, expectImage
-func HasCorrectResourceRequestsLimits(containerName string, cpuRequest, cpuLimit, memoryRequest, memoryLimit resource.Quantity) Predicate {
+// DeploymentContainerResourcesEqual verifies a reconciler deployment container
+// has the expected resource requests and limits.
+func DeploymentContainerResourcesEqual(expectedSpec v1beta1.ContainerResourcesSpec) Predicate {
 	return func(o client.Object) error {
 		if o == nil {
 			return ErrObjectNotFound
 		}
+		if uObj, ok := o.(*unstructured.Unstructured); ok {
+			rObj, err := kinds.ToTypedObject(uObj, core.Scheme)
+			if err != nil {
+				return err
+			}
+			o, err = kinds.ObjectAsClientObject(rObj)
+			if err != nil {
+				return err
+			}
+		}
 		dep, ok := o.(*appsv1.Deployment)
 		if !ok {
 			return WrongTypeErr(o, &appsv1.Deployment{})
 		}
-		container := ContainerByName(dep, containerName)
-		if container == nil {
-			return fmt.Errorf("expected container not found: %s", containerName)
-		}
-		if !equality.Semantic.DeepEqual(container.Resources.Requests[corev1.ResourceCPU], cpuRequest) {
-			return errors.Errorf("The CPU request of the %q container should be %v, got %v", container.Name, cpuRequest, container.Resources.Requests[corev1.ResourceCPU])
+		container := ContainerByName(dep, expectedSpec.ContainerName)
+		return validateContainerResources(container, expectedSpec)
+	}
+}
+
+// DeploymentContainerResourcesAllEqual verifies all reconciler deployment
+// containers have the expected resource requests and limits.
+func DeploymentContainerResourcesAllEqual(expectedByName map[string]v1beta1.ContainerResourcesSpec) Predicate {
+	return func(o client.Object) error {
+		if o == nil {
+			return ErrObjectNotFound
 		}
-		if !equality.Semantic.DeepEqual(container.Resources.Limits[corev1.ResourceCPU], cpuLimit) {
-			return errors.Errorf("The CPU limit of the %q container should be %v, got %v", container.Name, cpuLimit, container.Resources.Limits[corev1.ResourceCPU])
+		if uObj, ok := o.(*unstructured.Unstructured); ok {
+			rObj, err := kinds.ToTypedObject(uObj, core.Scheme)
+			if err != nil {
+				return err
+			}
+			o, err = kinds.ObjectAsClientObject(rObj)
+			if err != nil {
+				return err
+			}
 		}
-		if !equality.Semantic.DeepEqual(container.Resources.Requests[corev1.ResourceMemory], memoryRequest) {
-			return errors.Errorf("The memory request of the %q container should be %v, got %v", container.Name, memoryRequest, container.Resources.Requests[corev1.ResourceMemory])
+		d, ok := o.(*appsv1.Deployment)
+		if !ok {
+			return WrongTypeErr(d, &appsv1.Deployment{})
 		}
-		if !equality.Semantic.DeepEqual(container.Resources.Limits[corev1.ResourceMemory], memoryLimit) {
-			return errors.Errorf("The memory limit of the %q container should be %v, got %v", container.Name, memoryLimit, container.Resources.Limits[corev1.ResourceMemory])
+		for _, container := range d.Spec.Template.Spec.Containers {
```

Comment:
This would pass if the actual deployment has no containers (or fewer than the expected containers). Is that intentional?

Reply:
The default map includes all the possible containers, so we don't really want to require them all to match, otherwise we have to filter & copy the map being used for every assertion. 

I think it's probably fine to just test the resources and assume the right containers exist and are tested by other tests.

Reply:
Or do you think I should add a helper function to filter the map by a specific set of expected containers before passing it to the predicate?

Reply:
Yeah I think it would be a stronger assertion if we ensured the expected containers are in the deployment. I like the suggestion of a helper function or a parameter to filter by a list of expected containers

Reply:
added sorted container name comparison